### PR TITLE
solve_network refactor, ERM, & opts testing 

### PR DESCRIPTION
**Pull request by ktehranchi**
## Changes proposed in this Pull Request
- Refactors solve_network module to break opts into sub-modules which we import. Motivated by the unruly size the solve_network module had become. 
- Implements Energy Reserve Margin (ERM) opt. Follows implementation described in [Stephen & Kirschen 2024](https://ieeexplore.ieee.org/document/10667201)
- Implements test for ERM. I have added placeholders for other opts. The structure of the tests will reuse basic 3-bus test networks

## Checklist

- [x] I tested my contribution locally and it seems to work fine.
- [x] Code and workflow changes are sufficiently documented.
- [x] Changed dependencies are added to `envs/environment.yaml`.
- [ ] Changes in configuration options are added in all of `config.default.yaml`.
- [ ] Changes in configuration options are also documented in `doc/configtables/*.csv`.

### CODE REVIEW

**Comment by ktehranchi** [comment]
@trevorb1 I'm down to merge this in after the v0.6.0 release. I'm just requesting review to see if you have any feedback on how I have restructured the opts and solve_network. No need to do detailed review of the new ERM.

File: workflow/scripts/test/conftest.py [hunk 1]
```diff
@@ -0,0 +1,551 @@
...
+        "Carrier",
+        "battery",
+        co2_emissions=0,
+    )
+    n.add(
+        "Carrier",
+        "AC",
+        co2_emissions=0,
+    )
+    return n
```
Comment (trevorb1) [inline workflow/scripts/test/conftest.py:303]: I really like the `base_network()` fixture, but wonder too much is included in it? Carriers that are here, but only used in some constraints (ie. `nuclear`), component interactions that shouldnt happen are included (ie. lines and links both being used for power transfer), attributes that are only used for particular constraints are included globally (ie. `land_area`), ect. 

This can be an issue as then we are not isolating each test. For example, (a fictional case could be) if a constraint uses a shared function to check that component names are defined correctly and there is an issue with how a line is defined. All tests would fail cause lines are always included in the base network. However, if running a transport model, these tests should pass as lines are replaced with links. This can make diagnosing the problem harder. (I do concede that is is a fairly extreme edge case 😅)

Would it be better to have like a high level `base_network()` that has only components that are shared across all network configurations (ie. snapshots, buses, loads)? Then we can have a few "sub" base networks. For example, one that is transport or not transport; one that is electrical only vs sector (where, for example some generators are replaced with links); ect? 

Or maybe the base network is literally a 1, 2, or 3 node network that only defines the snapshots and buses. Then each constraint adds in exactly what it needs? ie. for some constraints, we dont even need multi-node networks, or storage units? 

Or is this creating too much structure?

File: workflow/scripts/test/conftest.py [hunk 2]
```diff
@@ -0,0 +1,551 @@
...
+            )
+
+    snakemake = MockSnakemake()
+
+    yield config, snakemake
+
+    # Clean up temporary files
+    os.unlink(portfolio_standards_file.name)
+    os.unlink(rps_reeds_file.name)
+    os.unlink(ces_reeds_file.name)
```
Comment (trevorb1) [inline workflow/scripts/test/conftest.py:464]: Would it be easier to just have simple CSVs in a fixtures folder, as that how its actually used in the workflow? Not that there is anything wrong with your implementation! Would just get around having to do the `NamedTemporaryFile` I think?

File: workflow/scripts/test/test_content.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+    )
+
+    n.add(
+        "Load",
+        "load_east",
+        bus="bus3",
+        p_set=pd.Series(250, index=n.snapshots),
+    )
+
+    return n
```
Comment (trevorb1) [inline workflow/scripts/test/test_content.py:244]: Why cant you use the base network here, and build onto it like in the other tests?

File: workflow/scripts/test/test_content.py [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+            )
+
+    snakemake = MockSnakemake()
+
+    yield config, snakemake
+
+    # Clean up temporary files
+    os.unlink(portfolio_standards_file.name)
+    os.unlink(rps_reeds_file.name)
+    os.unlink(ces_reeds_file.name)
```
Comment (trevorb1) [inline workflow/scripts/test/test_content.py:322]: Wasnt something similar already defined in conftest?

File: workflow/scripts/test/README.md [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+  - Data storage
+  - Storage contribution to ERM
+  - Transmission line contribution to ERM
+
+- Tests for ERM constraints across multiple regions
+  - Constraint application in multiple regions
+  - Interconnection contribution to ERM
+
+- Tests for basic functionality of PRM constraints
+  - Constraint creation
```
Comment (trevorb1) [inline workflow/scripts/test/README.md:29]: Do we actually test the constraint creation? I may have missed it, but seems like we test constraint behaviour? I would consider testing constraint creation to be how linopy describes [here](https://linopy.readthedocs.io/en/latest/testing-framework.html#Testing-with-Linopy). (Sorry, this one is very nitpicky!)

**Review by trevorb1 — COMMENTED** [review]
Hi @ktehranchi; this looks really really great! Thanks for staring this (long overdue) testing setup. Please see below for my comments. Also, a couple general questions: 

1. Im not sure why we need the `sys.path.append(os.path.join(os.path.dirname(__file__), ".."))` at the top of all the testing files? 
2. Are we wanting to add the tests to the GitHub Workflow as well? Or you think since each test does an optimization (ie, its kinda heavy for a test) we leave it up to the user to test locally? 
3. We should add that users should run tests in the contributing guide in the docs as well? (ie. the code snippet you have in the README.md copy/paste to the contributing guide)

File: workflow/scripts/test/README.md (hunk 5 above)
Comment (ktehranchi) [reply workflow/scripts/test/README.md:29]: I used to test the constraint creation, so I should get rid of this from the README too.

File: workflow/scripts/test/conftest.py (hunk 2 above)
Comment (ktehranchi) [reply workflow/scripts/test/conftest.py:464]: thats true, I will do that with some test data in the same format

File: workflow/scripts/test/conftest.py (hunk 1 above)
Comment (ktehranchi) [reply workflow/scripts/test/conftest.py:303]: I like this idea, to have a high level base_network then some sub networks that are made to test different possible configurations.... I'll make this change, thanks!

File: workflow/scripts/solve_network.py [hunk 6]
```diff
@@ -25,133 +25,47 @@
...
+from opts.sector import (
+    add_cooling_heat_pump_constraints,
+    add_demand_response_constraint,
+    add_ev_generation_constraint,
+    add_gshp_capacity_constraint,
+    add_ng_import_export_limits,
+    add_sector_co2_constraints,
+    add_sector_demand_response_constraints,
+    add_water_heater_constraints,
+)
```
[the remark this answers is not in the document]
Comment (ktehranchi) [reply workflow/scripts/solve_network.py:59]: I don't think we do since these are just relative imports??

File: workflow/scripts/test/README.md
[the remark this answers is not in the document]
Comment (ktehranchi) [reply workflow/scripts/test/README.md:1]: oh yeah I guess normally the test would go in the base directory. I am going to leave restructuring the workflow for packaging for another PR. We had mentioned this previously in  #150

File: workflow/scripts/test/run_tests.py
[the remark this answers is not in the document]
Comment (ktehranchi) [reply workflow/scripts/test/run_tests.py:1]: removed!

File: workflow/scripts/test/README.md [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+## Requirements
+
+- **pytest**: The tests use pytest as the testing framework
+- **GLPK**: The tests use the GLPK solver by default
+
+Install the required packages:
+
+```bash
+pip install pytest
+```
```
[the remark this answers is not in the document]
Comment (ktehranchi) [reply workflow/scripts/test/README.md:98]: Removed this from readme, and added pytest to dev.yaml. It was already in `uv`

File: workflow/scripts/opts/sector.py
[the remark this answers is not in the document]
Comment (ktehranchi) [reply workflow/scripts/opts/sector.py:1]: yep direct copy and paste

File: workflow/scripts/solve_network.py [hunk 8]
```diff
@@ -1742,6 +230,76 @@ def run_optimize(n, rolling_horizon, skip_iterations, cf_solving, **kwargs):
...
+                n.add(nm, df_idx, **df.loc[df_idx])
+        logger.info(n.consistency_check())
+
+        # copy time-dependent
+        selection = n.component_attrs[nm].type.str.contains("series")
+        for tattr in n.component_attrs[nm].index[selection]:
+            n.import_series_from_dataframe(time_df[tattr], nm, tattr)
+
+    # roll over the last snapshot of time varying storage state of charge to be the state_of_charge_initial for the next time period
+    n.storage_units.loc[:, "state_of_charge_initial"] = n.storage_units_t.state_of_charge.loc[planning_horizon].iloc[-1]
```
[the remark this answers is not in the document]
Comment (ktehranchi) [reply workflow/scripts/solve_network.py:300]: This is just refactored/functionalized from the recent #518 !

**Comment by ktehranchi** [comment]
> Hi @ktehranchi; this looks really really great! Thanks for staring this (long overdue) testing setup. Please see below for my comments. Also, a couple general questions:
> 
> 1. Im not sure why we need the `sys.path.append(os.path.join(os.path.dirname(__file__), ".."))` at the top of all the testing files?

This has to do with your comment on the packaging and structure of the repository. im going to leave this to address in #150 

> 2. Are we wanting to add the tests to the GitHub Workflow as well? Or you think since each test does an optimization (ie, it's kinda heavy for a test) we leave it up to the user to test locally?

I don't think the optimization is too heavy! the test networks are all <48 hours of data, and glpk can solve it extremely quickly. We can make it even lighter weight. I will try to set up the online testing at some point per #207 

> 3. We should add that users should run tests in the contributing guide in the docs as well? (ie. the code snippet you have in the README.md copy/paste to the contributing guide)

I will add this

**Comment by ktehranchi** [comment]
Also thank you for the super detailed code review!!! @trevorb1

File: workflow/scripts/test/conftest.py (hunk 1 above)
Comment (ktehranchi) [reply workflow/scripts/test/conftest.py:303]: created in #598
